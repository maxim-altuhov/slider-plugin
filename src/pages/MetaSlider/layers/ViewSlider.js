import $ from 'jquery';
import Observer from '../patterns/Observer';

class ViewSlider extends Observer {
  constructor() {
    super();
    this.isFirstInit = true;
  }

  init(options) {
    this.$elemSlider = options.$elemSlider;
    this.$sliderProgress = options.$sliderProgress;
    this.$elemThumbs = options.$elemThumbs;
    this.$elemMarkers = options.$elemMarkers;
    this.setEventsSlider();
  }

  update(options) {
    if (this.isFirstInit) {
      this.init(options);
      this.isFirstInit = false;
    }

    const { key } = options;
    const setValueVerifKeys = (
      key === 'init'
      || key === 'showBackground'
      || key === 'mainColor'
      || key === 'changedValue'
      || key === 'initValueFirst'
      || key === 'initValueSecond'
      || key === 'isRange'
      || key === 'minValue'
      || key === 'maxValue'
      || key === 'customValues'
    );
    const autoMarginVerifKeys = (
      key === 'init'
      || key === 'initAutoMargins'
      || key === 'showMarkers'
      || key === 'showScale'
      || key === 'isVertical'
    );
    const minAndMaxVerifKeys = (
      key === 'init'
      || key === 'minValue'
      || key === 'maxValue'
      || key === 'customValues'
    );

    if (setValueVerifKeys) this.setBackgroundTheRange(options);
    if (key === 'secondColor' || key === 'init') this.setBackgroundForSlider(options);
    if (key === 'isVertical' || key === 'init') this.setVerticalOrientation(options);
    if (autoMarginVerifKeys) this.setAutoMargins(options);
    if (minAndMaxVerifKeys) this.setMinAndMaxVal(options);
  }

  renderSlider(initSelector) {
    if (!this.$selector) this.$selector = initSelector;

    const $fragmentWithASlider = $(document.createDocumentFragment());
    const $blockSlider = $(document.createElement('div'));
    $blockSlider.addClass('meta-slider js-meta-slider');

    const HTMLBlock = `<div class="meta-slider__progress js-meta-slider__progress"></div>
    <button type="button" class="meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_left" data-value="">
      <span class="meta-slider__marker js-meta-slider__marker meta-slider__marker_left"></span>
    </button>
    <button type="button" class="meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_right" data-value="">
      <span class="meta-slider__marker js-meta-slider__marker meta-slider__marker_right"></span>
    </button>
    <div class="meta-slider__scale js-meta-slider__scale"></div>`;

    $blockSlider.html(HTMLBlock);

    $fragmentWithASlider.append($blockSlider);
    this.$selector.append($fragmentWithASlider);
  }

  setMinAndMaxVal(options) {
    const {
      minValue,
      maxValue,
      customValues,
    } = options;

    this.$elemSlider.attr({ 'data-min': minValue, 'data-max': maxValue });

    if (customValues.length > 0) {
      const firstCustomElem = customValues[0];
      const lastCustomElem = customValues[customValues.length - 1];
      this.$elemSlider.attr({ 'data-min_text': firstCustomElem, 'data-max_text': lastCustomElem });
    } else {
      this.$elemSlider.removeAttr('data-min_text data-max_text');
    }
  }

  setAutoMargins(options) {
    const {
      initAutoMargins,
      showMarkers,
      showScale,
      isVertical,
    } = options;
    const verifiableProp = (initAutoMargins && !isVertical);

    if (verifiableProp && showMarkers) {
      const heightMarker = this.$elemMarkers.eq(-1).outerHeight();
      const heightThumb = this.$elemThumbs.eq(-1).outerHeight();

      this.$elemSlider.css('margin-top', `${heightMarker + (heightThumb / 1.5)}px`);
    } else {
      this.$elemSlider.css('margin-top', '');
    }

    if (verifiableProp && showScale) {
      const elemScalePoints = this.$selector.find('.js-meta-slider__scale-point');
      this.$elemSlider.css('margin-bottom', `${elemScalePoints.eq(0).outerHeight() * 3}px`);
    } else {
      this.$elemSlider.css('margin-bottom', '');
    }
  }

  setBackgroundForSlider(options) {
    this.$elemSlider.css('background-color', options.secondColor);
  }

  setVerticalOrientation(options) {
    if (options.isVertical) {
      this.$elemSlider.addClass('meta-slider_vertical');
      this.$selector.addClass('selector-vertical');
    } else {
      this.$elemSlider.removeClass('meta-slider_vertical');
      this.$selector.removeClass('selector-vertical');
    }
  }

  setBackgroundTheRange(options) {
    if (options.showBackground) {
      const { valuesAsPercentageArray, mainColor } = options;
      const [firstPosition, secondPosition] = valuesAsPercentageArray;
      const settingForRange = { left: `${firstPosition}%`, right: `${100 - secondPosition}%`, background: mainColor };

      this.$sliderProgress.css(settingForRange);
    } else {
      this.$sliderProgress.css('background', 'none');
    }
  }

  handleSetSliderValues(event) {
    event.preventDefault();
    const $target = $(event.target);

    if ($target.hasClass('js-meta-slider')) {
      this.notify(event);
    }
  }

  setEventsSlider() {
    this.$elemSlider.on('pointerdown.slider', this.handleSetSliderValues.bind(this));
  }
}

export default ViewSlider;
