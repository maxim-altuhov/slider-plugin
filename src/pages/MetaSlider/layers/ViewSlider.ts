import $ from 'jquery';
import Observer from '../patterns/Observer';

class ViewSlider extends Observer {
  isFirstInit: boolean;
  $selector: JQuery;
  $elemSlider: JQuery;
  $sliderProgress: JQuery;
  $elemThumbs: JQuery;
  $elemMarkers: JQuery;

  constructor() {
    super();
    this.isFirstInit = true;
  }

  // Обновление view
  update(options: IPluginOptions) {
    if (this.isFirstInit) {
      this._init(options);
      this.isFirstInit = false;
    }

    const { key } = options;

    // prettier-ignore
    const setValueVerifKeys = (
      key === 'init' ||
      key === 'showBackground' ||
      key === 'mainColor' ||
      key === 'changedValue' ||
      key === 'initValueFirst' ||
      key === 'initValueSecond' ||
      key === 'isRange' ||
      key === 'minValue' ||
      key === 'maxValue' ||
      key === 'customValues' ||
      key === 'step' ||
      key === 'calcNumberOfDecimalPlaces' ||
      key === 'numberOfDecimalPlaces'
    );

    // prettier-ignore
    const autoMarginVerifKeys = (
      key === 'init' ||
      key === 'initAutoMargins' ||
      key === 'showMarkers' ||
      key === 'showScale' ||
      key === 'isVertical'
    );

    // prettier-ignore
    const minAndMaxVerifKeys = (
      key === 'init' ||
      key === 'minValue' ||
      key === 'maxValue' ||
      key === 'customValues'
    );

    if (setValueVerifKeys) this._setBackgroundTheRange(options);
    if (key === 'secondColor' || key === 'init') this._setBackgroundForSlider(options);
    if (key === 'isVertical' || key === 'init') this._setVerticalOrientation(options);
    if (autoMarginVerifKeys) this._setAutoMargins(options);
    if (minAndMaxVerifKeys) this._setMinAndMaxVal(options);
  }

  // Первоначальный рендер слайдера
  renderSlider(initSelector: JQuery) {
    if (!this.$selector) this.$selector = initSelector;

    const $fragmentWithASlider = $(document.createDocumentFragment());
    const $blockSlider = $(document.createElement('div'));
    $blockSlider.addClass('meta-slider js-meta-slider');

    const HTMLBlock = `<div class="meta-slider__progress js-meta-slider__progress"></div>
    <button type="button" class="meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_left" data-value="">
      <span class="meta-slider__marker js-meta-slider__marker meta-slider__marker_left"></span>
    </button>
    <button type="button" class="meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_right" data-value="">
      <span class="meta-slider__marker js-meta-slider__marker meta-slider__marker_right"></span>
    </button>
    <div class="meta-slider__scale js-meta-slider__scale"></div>`;

    $blockSlider.html(HTMLBlock);

    $fragmentWithASlider.append($blockSlider);
    this.$selector.append($fragmentWithASlider);
  }

  // Первоначальная инициализация
  private _init(options: IPluginOptions) {
    this.$elemSlider = options.$elemSlider;
    this.$sliderProgress = options.$sliderProgress;
    this.$elemThumbs = options.$elemThumbs;
    this.$elemMarkers = options.$elemMarkers;
    this._setEventsSlider();
  }

  // Устанвливает data-атрибуты с мин. и макс. значениями слайдера
  private _setMinAndMaxVal(options: IPluginOptions) {
    const { minValue, maxValue, customValues } = options;

    this.$elemSlider.attr({ 'data-min': minValue, 'data-max': maxValue });

    if (customValues.length > 0) {
      const firstCustomElem = customValues[0];
      const lastCustomElem = customValues[customValues.length - 1];
      this.$elemSlider.attr({ 'data-min_text': firstCustomElem, 'data-max_text': lastCustomElem });
    } else {
      this.$elemSlider.removeAttr('data-min_text data-max_text');
    }
  }

  // Проверка и установка отступов у слайдера
  private _setAutoMargins(options: IPluginOptions) {
    // prettier-ignore
    const { 
      initAutoMargins,
      showMarkers,
      showScale,
      isVertical,
    } = options;

    const verifProp = initAutoMargins && !isVertical;

    if (verifProp && showMarkers) {
      const heightMarker = this.$elemMarkers.eq(-1).outerHeight();
      const heightThumb = this.$elemThumbs.eq(-1).outerHeight();

      this.$elemSlider.css('margin-top', `${heightMarker + heightThumb / 1.5}px`);
    } else {
      this.$elemSlider.css('margin-top', '');
    }

    if (verifProp && showScale) {
      const elemScalePoints = this.$selector.find('.js-meta-slider__scale-point');
      this.$elemSlider.css('margin-bottom', `${elemScalePoints.eq(0).outerHeight() * 3}px`);
    } else {
      this.$elemSlider.css('margin-bottom', '');
    }
  }

  // Установка фонового цвета для слайдера
  private _setBackgroundForSlider(options: IPluginOptions) {
    this.$elemSlider.css('background-color', options.secondColor);
  }

  // Сброс или установка вертикальной ориентации слайдера
  private _setVerticalOrientation(options: IPluginOptions) {
    if (options.isVertical) {
      this.$elemSlider.addClass('meta-slider_vertical');
      this.$selector.addClass('ms-vertical');
    } else {
      this.$elemSlider.removeClass('meta-slider_vertical');
      this.$selector.removeClass('ms-vertical');
    }
  }

  // Проверка и установка цвета заливки интервала между бегунками
  private _setBackgroundTheRange(options: IPluginOptions) {
    if (options.showBackground) {
      const { valuesAsPercentageArray, mainColor } = options;
      const [firstPosition, secondPosition] = valuesAsPercentageArray;

      const settingForRange = {
        left: `${firstPosition}%`,
        right: `${100 - secondPosition}%`,
        background: mainColor,
      };

      this.$sliderProgress.css(settingForRange);
    } else {
      this.$sliderProgress.css('background', 'none');
    }
  }

  // Обработчик события клика внутри слайдера
  private _setEventsSlider() {
    this.$elemSlider.on('pointerdown.slider', this._handleSetSliderValues.bind(this));
  }

  // Получает значения при клике внутри слайдера
  private _handleSetSliderValues(event: IEvent) {
    event.preventDefault();
    const $target = $(event.target);

    if ($target.hasClass('js-meta-slider')) {
      this.notify(event);
    }
  }
}

export default ViewSlider;
