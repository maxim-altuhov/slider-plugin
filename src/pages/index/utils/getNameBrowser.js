function getNameBrowser() {
  const ua = navigator.userAgent;

  if (ua.search(/rv:11.0/) > 0) {
    return 'Internet Explorer 11';
  }
  if (ua.search(/MSIE/) > 0) {
    return 'Internet Explorer';
  }
  return 'ะะต Internet Explorer';
}

export default getNameBrowser;
