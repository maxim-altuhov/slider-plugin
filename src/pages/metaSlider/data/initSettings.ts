const initSettings: IPluginOptions = {
  key: '',
  $selector: $(),
  $elemSlider: $(),
  $sliderProgress: $(),
  $elemMarkers: $(),
  $elemScale: $(),
  $elemThumbs: $(),
  mainColor: '#6d6dff',
  secondColor: '#e4e4e4',
  colorMarker: '',
  colorThumb: '',
  colorTextForMarker: '#ffffff',
  colorBorderForMarker: '#ffffff',
  colorBorderForThumb: '#ffffff',
  colorForScale: '#000000',
  initFormatted: true,
  initAutoMargins: true,
  initScaleAdjustment: true,
  calcNumberOfDecimalPlaces: true,
  showError: false,
  showScale: true,
  showMarkers: true,
  showBackground: true,
  isRange: true,
  isVertical: false,
  initAutoScaleCreation: true,
  checkingStepSizeForScale: false,
  step: 1,
  stepAsPercent: 0,
  minValue: 0,
  maxValue: 0,
  stepSizeForScale: null,
  numberOfDecimalPlaces: 0,
  preFix: '',
  postFix: '',
  customValues: [],
  initValueFirst: null,
  initValueSecond: null,
  initValuesArray: [],
  textValueFirst: '',
  textValueSecond: '',
  textValuesArray: [],
  valuesAsPercentageArray: [],
};

export default initSettings;
