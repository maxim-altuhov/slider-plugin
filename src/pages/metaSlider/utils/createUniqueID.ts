function createUniqueID(): string {
  const IDPartOne = Date.now().toString(36);
  const IDPartTwo = Math.random().toString(36).substring(2);

  return IDPartOne + IDPartTwo;
}

export default createUniqueID;
