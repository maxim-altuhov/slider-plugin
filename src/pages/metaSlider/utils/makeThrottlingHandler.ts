// The any type is used, since the called functions can contain any arguments
function makeThrottlingHandler(fn: (...args: any[]) => unknown, timeout: number) {
  let timer: NodeJS.Timeout | null = null;

  return (...args: any[]) => {
    if (timer) return;

    timer = setTimeout(() => {
      fn(...args);

      if (timer) {
        clearTimeout(timer);
        timer = null;
      }
    }, timeout);
  };
}

export default makeThrottlingHandler;
