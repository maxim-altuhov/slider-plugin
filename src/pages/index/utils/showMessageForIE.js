import getNameBrowser from './getNameBrowser';

function showMessageForIE() {
  const currentBrowser = getNameBrowser();
  const thisIsIEBrowser = (currentBrowser === 'Internet Explorer 11') || (currentBrowser === 'Internet Explorer');

  if (thisIsIEBrowser) {
    document.body.innerHTML = 'Internet Explorer is no longer supported, please switch to a more modern browser!';
  }
}

export default showMessageForIE;
