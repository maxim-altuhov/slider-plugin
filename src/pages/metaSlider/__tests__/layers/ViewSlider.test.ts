/* eslint-disable @typescript-eslint/dot-notation */
import ViewSlider from '../../layers/ViewSlider';
import initSettings from '../../data/initSettings';
import * as createUniqueID from '../../utils/createUniqueID';

const classViewSlider = new ViewSlider();
const initSelectorName = 'render-selector';
const scaleSelectorName = '.js-meta-slider__scale';
const thumbSelectorName = '.js-meta-slider__thumb';
const markerSelectorName = '.js-meta-slider__marker';
const TEST_HEIGHT_ELEM = 20;

const initHTMLBlock = `<div class="slider-block" id="${initSelectorName}"></div>`;
const scalePointHTML = `<button type="button" class="meta-slider__scale-point
js-meta-slider__scale-point" style="height:${TEST_HEIGHT_ELEM}px"></button>`;
document.body.innerHTML = initHTMLBlock;
const $initSelector = $(`#${initSelectorName}`);

classViewSlider.renderSlider($initSelector);
$(scaleSelectorName).append(scalePointHTML);

$(thumbSelectorName).eq(-1).css('height', TEST_HEIGHT_ELEM);
$(markerSelectorName).eq(-1).css('height', TEST_HEIGHT_ELEM);

describe('Checking the "ViewSlider" layer. State before first initialization the "renderSlider" and "update" method', () => {
  const notInitViewSlider = new ViewSlider();

  test('State before first initialization', () => {
    expect($initSelector.html()).toMatchSnapshot();
    expect(notInitViewSlider.observerList).toHaveLength(0);
    expect(notInitViewSlider['_$selector']).toHaveLength(0);
    expect(notInitViewSlider['_$elemSlider']).toHaveLength(0);
    expect(notInitViewSlider['_$sliderProgress']).toHaveLength(0);
    expect(notInitViewSlider['_$elemThumbs']).toHaveLength(0);
    expect(notInitViewSlider['_$elemMarkers']).toHaveLength(0);
    expect(notInitViewSlider['_isFirstInit']).toBe(true);
  });
});

describe('Checking the "ViewSlider" layer', () => {
  const defaultSettings = {
    key: 'init',
    $selector: $(`#${initSelectorName}`),
    $elemSlider: $('.js-meta-slider'),
    $sliderProgress: $('.js-meta-slider__progress'),
    $elemThumbs: $('.js-meta-slider__thumb'),
    $elemMarkers: $('.js-meta-slider__marker'),
    $elemScale: $('.js-meta-slider__scale'),
    showBackground: true,
    initAutoMargins: true,
    showMarkers: true,
    showScale: true,
    isVertical: false,
    mainColor: 'red',
    secondColor: 'grey',
    minValue: 0,
    maxValue: 100,
    valuesAsPercentageArray: [0, 100],
    customValues: [],
  };
  let testSettings: IPluginOptions;
  let $selector: JQuery<HTMLElement>;
  let $elemSlider: JQuery<HTMLElement>;
  let $sliderProgress: JQuery<HTMLElement>;

  beforeAll(() => {
    testSettings = $.extend({}, initSettings, defaultSettings);
    classViewSlider.update(testSettings);

    $selector = classViewSlider['_$selector'];
    $elemSlider = classViewSlider['_$elemSlider'];
    $sliderProgress = classViewSlider['_$sliderProgress'];
  });

  afterEach(() => {
    testSettings = $.extend({}, initSettings, defaultSettings);
  });

  test('Checking the "renderSlider" method', () => {
    const mockCreateUniqueID = jest.spyOn(createUniqueID, 'default').mockImplementation(() => '');
    const checkingSelectorsArr = [
      /js-meta-slider/,
      /js-meta-slider__progress/,
      /js-meta-slider__thumb/,
      /js-meta-slider__marker/,
      /js-meta-slider__scale/,
      /data-id/,
      /data-value/,
      /data-text/,
    ];

    classViewSlider.renderSlider($initSelector);

    checkingSelectorsArr.forEach((selector) => {
      expect(document.body.innerHTML).toMatch(selector);
    });
    expect(mockCreateUniqueID).toHaveBeenCalled();
    expect(document.body.innerHTML).not.toBe(initHTMLBlock);
    expect($selector).toHaveLength(1);
    expect($selector).toBe($initSelector);
  });

  test('Checking the "_init" method', () => {
    jest.spyOn<ViewSlider, any>(classViewSlider, '_setEventsSlider');

    classViewSlider['_isFirstInit'] = true;
    classViewSlider.update(testSettings);

    expect(classViewSlider['_isFirstInit']).toBe(false);
    expect(classViewSlider['_$elemSlider']).toHaveLength(1);
    expect(classViewSlider['_$sliderProgress']).toHaveLength(1);
    expect(classViewSlider['_$elemThumbs']).toHaveLength(2);
    expect(classViewSlider['_$elemMarkers']).toHaveLength(2);
    expect(classViewSlider['_setEventsSlider']).toHaveBeenCalled();
  });

  test('Checking the "_setMinAndMaxVal" method => "defaultSettings"', () => {
    classViewSlider.update(testSettings);
    const { minValue, maxValue } = testSettings;

    expect($elemSlider.attr('data-min')).toBe(String(minValue));
    expect($elemSlider.attr('data-max')).toBe(String(maxValue));
    expect($elemSlider.attr('data-min_text')).toBeUndefined();
    expect($elemSlider.attr('data-max_text')).toBeUndefined();
  });

  test('Checking the "_setMinAndMaxVal" method => option "customValues"', () => {
    testSettings.key = 'customValues';
    testSettings.customValues = ['one', 'two', 'three'];
    classViewSlider.update(testSettings);

    const { customValues } = testSettings;
    expect($elemSlider.attr('data-min_text')).toBe(String(customValues[0]));
    expect($elemSlider.attr('data-max_text')).toBe(String(customValues[customValues.length - 1]));

    testSettings.key = 'customValues';
    testSettings.customValues = [];
    classViewSlider.update(testSettings);

    expect($elemSlider.attr('data-min_text')).toBeUndefined();
    expect($elemSlider.attr('data-max_text')).toBeUndefined();
  });

  test('Checking the "_setAutoMargins" method => "defaultSettings"', () => {
    classViewSlider.update(testSettings);

    expect($elemSlider.css('margin-top')).toMatch(/\d+px/);
    expect($elemSlider.css('margin-top')).not.toBe('0px');
    expect($elemSlider.css('margin-bottom')).toMatch(/\d+px/);
    expect($elemSlider.css('margin-bottom')).not.toBe('0px');
  });

  test('Checking the "_setAutoMargins" method => option "showScale"', () => {
    testSettings.key = 'showScale';
    testSettings.showScale = false;
    classViewSlider.update(testSettings);

    expect($elemSlider.css('margin-bottom')).toBe('');
  });

  test('Checking the "_setAutoMargins" method => option "showMarkers"', () => {
    testSettings.key = 'showMarkers';
    testSettings.showMarkers = false;
    classViewSlider.update(testSettings);

    expect($elemSlider.css('margin-top')).toBe('');
  });

  test('Checking the "_setAutoMargins" method => option "initAutoMargins"', () => {
    testSettings.key = 'initAutoMargins';
    testSettings.initAutoMargins = false;
    classViewSlider.update(testSettings);

    expect($elemSlider.css('margin-top')).toBe('');
    expect($elemSlider.css('margin-bottom')).toBe('');
  });

  test('Checking the "_setAutoMargins" method => option "isVertical"', () => {
    testSettings.key = 'isVertical';
    testSettings.isVertical = true;
    classViewSlider.update(testSettings);

    expect($elemSlider.css('margin-top')).toBe('');
    expect($elemSlider.css('margin-bottom')).toBe('');
  });

  test('Checking the "_setBackgroundForSlider" method => "defaultSettings"', () => {
    classViewSlider.update(testSettings);
    const { secondColor } = testSettings;

    expect($elemSlider.prop('style')).toHaveProperty('background-color', secondColor);
  });

  test('Checking the "_setBackgroundForSlider" method => option "secondColor"', () => {
    testSettings.key = 'secondColor';
    testSettings.secondColor = 'yellow';
    classViewSlider.update(testSettings);
    const { secondColor } = testSettings;

    expect($elemSlider.prop('style')).toHaveProperty('background-color', secondColor);
  });

  test('Checking the "_setVerticalOrientation" method => "defaultSettings"', () => {
    classViewSlider.update(testSettings);

    expect($selector.attr('class')).not.toMatch(/ms-vertical/);
    expect($elemSlider.attr('class')).not.toMatch(/meta-slider_vertical/);
  });

  test('Checking the "_setVerticalOrientation" method => option "isVertical"', () => {
    testSettings.key = 'isVertical';
    testSettings.isVertical = true;
    classViewSlider.update(testSettings);

    expect($selector.attr('class')).toMatch(/ms-vertical/);
    expect($elemSlider.attr('class')).toMatch(/meta-slider_vertical/);
  });

  test('Checking the "_setBackgroundTheRange" method => "defaultSettings"', () => {
    classViewSlider.update(testSettings);
    const { valuesAsPercentageArray, mainColor } = testSettings;
    const [firstPosition, secondPosition] = valuesAsPercentageArray;

    expect($sliderProgress.prop('style')).toHaveProperty('left', `${firstPosition}%`);
    expect($sliderProgress.prop('style')).toHaveProperty('right', `${100 - secondPosition}%`);
    expect($sliderProgress.prop('style')).toHaveProperty('background', mainColor);
  });

  test('Checking the "_setBackgroundTheRange" method => option "showBackground"', () => {
    testSettings.key = 'showBackground';
    testSettings.showBackground = false;
    classViewSlider.update(testSettings);

    expect($sliderProgress.prop('style')).toHaveProperty('background', 'none');
  });

  test('Checking the "_setEventsSlider" method, event "pointerdown" => _handleSetSliderValues', () => {
    const mockNotify = jest.spyOn(classViewSlider, 'notify');
    const eventPointerdown = $.Event('pointerdown.slider');
    eventPointerdown.preventDefault = jest.fn();

    $elemSlider.trigger(eventPointerdown);

    expect(eventPointerdown.preventDefault).toHaveBeenCalled();
    expect(mockNotify).toHaveBeenCalledWith(eventPointerdown);
  });
});
