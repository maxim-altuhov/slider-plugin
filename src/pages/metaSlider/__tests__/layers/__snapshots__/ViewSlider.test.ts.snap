// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checking the "ViewSlider" layer. State before first initialization the "renderSlider" and "update" method State before first initialization 1`] = `
"<div class=\\"meta-slider js-meta-slider\\"><div class=\\"meta-slider__progress js-meta-slider__progress\\"></div>
    <button type=\\"button\\" class=\\"meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_left\\" data-value=\\"\\" data-text=\\"\\">
      <span class=\\"meta-slider__marker js-meta-slider__marker meta-slider__marker_left\\"></span>
    </button>
    <button type=\\"button\\" class=\\"meta-slider__thumb js-meta-slider__thumb meta-slider__thumb_right\\" data-value=\\"\\" data-text=\\"\\" style=\\"height: 20px;\\">
      <span class=\\"meta-slider__marker js-meta-slider__marker meta-slider__marker_right\\" style=\\"height: 20px;\\"></span>
    </button>
    <div class=\\"meta-slider__scale js-meta-slider__scale\\"><button type=\\"button\\" class=\\"meta-slider__scale-point
js-meta-slider__scale-point\\" style=\\"height:20px\\"></button></div></div>"
`;
